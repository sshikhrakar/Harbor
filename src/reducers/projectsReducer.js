import {
  DOWNLOAD_STARTED,
  DOWNLOAD_COMPLETED,
  FETCH_ALL_PROJECTS_FULFILLED,
} from '../actions/actionTypes';

export const initialState = {};

/**
 * Reducer for projects.
 *
 * @param {Object} state=initialState
 * @param {Object} action
 * @returns {Object}
 */
function projectsReducer(state = initialState, action) {
  switch(action.type) {

    case FETCH_ALL_PROJECTS_FULFILLED:
      return {
        ...action.payload.projects,
      };

    case DOWNLOAD_COMPLETED:
      return {
        ...state,
        [action.payload.project.packageName]: projectReducer(state[action.payload.project.packageName], action),
      };

    default:
      return state;
  }
}

/**
 * Reducer for a single project.
 *
 * @param {Object} state
 * @param {Object} action
 * @returns {Object}
 */
function projectReducer(state, action) {
  switch(action.type) {

    case DOWNLOAD_STARTED:
      return {
        ...state,
        uploads: {
          ...state.uploads,
          [action.payload.timestamp]: {
            ...state.uploads[action.payload.timestamp],
            downloading: true,
          },
        },
      };


    case DOWNLOAD_COMPLETED:
      return {
        ...state,
        uploads: {
          ...state.uploads,
          [action.payload.timestamp]: {
            ...state.uploads[action.payload.timestamp],
            downloading: false,
            downloaded: true,
            apkPath: action.payload.apkPath,
          },
        },
      };

    default:
      return state;
  }
}

export default projectsReducer;
