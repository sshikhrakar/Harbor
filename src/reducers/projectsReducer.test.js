import projectsReducer, { initialState } from './projectsReducer';
import {
  fetchAllProjectsFulfilled,
} from '../actions/projectActions';

describe('REDUCER: PROJECTS', () => {

  it('should render the initial state when state is undefined and action is an empty object', () => {
    expect(projectsReducer(undefined, {})).toEqual(initialState);
  });

  it('should set projects key when FETCH_ALL_PROJECTS_FULFILLED is received', () => {
    expect(projectsReducer(fetchAllProjectsFulfilled('data'), {})).toMatchSnapshot();
  });
});
