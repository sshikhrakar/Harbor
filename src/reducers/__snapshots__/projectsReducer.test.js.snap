// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REDUCER: PROJECTS should set projects key when FETCH_ALL_PROJECTS_FULFILLED is received 1`] = `
Object {
  "payload": Object {
    "projects": "data",
  },
  "type": "PROJECTS/FETCH_ALL_PROJECTS_FULFILLED",
}
`;
